### DISCUSSION TITLE
Window of AU plugin not resizable

### DISCUSSION THREAD

**Discussion by bijlevel** [RomanKubiak/ctrlr#336] [Q&A]
Maybe already discussed on the old forum (if so please point me to the discussion), but when I export a restricted instance as AU component, the window of the component is not resizable. This is rather annoying, especially when the original size of panel is rather big. Is there a solution for this problem?

**Comment by damiensellier** [comment]
Are you talking about changing the zoom factor and the wrapper window not resizing to the content?

**Reply by bijlevel** [reply]
Yes (and resizing with the zooming of the content should be ideal!), but even "manual" resizing is not possible with the plugin. I guess it is some wrong setting when exporting, so maybe looking into the source code of CTRLR is an option, but I don't have a clue where to start looking!

**Reply by damiensellier** [reply]
To my knowledge, it's not possible to resize a vst or AU instance compared to the standalone app that can zoom in/out because the way plugins are constructed in the host, once the frame/container is set you cannot change it.
I saw some plugin zoom in and out with the proper container size resizing like the ones from Arturia, but I have no idea if it's possible with ctrlr/JUCE. The only thread in the old forum about zoom is this one : 
[export restricted instance + zoom panel problem](https://ctrlr.org/forums/topic/export-restricted-instance-hidden-menu-bar/)
The trick Mr.Tor gave in [this post](https://ctrlr.org/forums/topic/export-restricted-instance-hidden-menu-bar/#post-72363) is gold because it's the only way to resize a vst/au instance. I use it everytime.

Goodweather got deeper in the zoom feature for restricted instances on this very post : 
["I solved the zoom and window size issue for restricted instance"](https://ctrlr.org/forums/topic/export-restricted-instance-hidden-menu-bar/#post-98976)

I worked on a plugin recently and exported 2 instances for 2 sizes (HD ratio 1.0, and, HiDPI screen ratio 2.0) because of that.

**Reply by bijlevel** [reply]
Thank you so much for your input! It seems indeed an existing problem, not yet solved. What I understand from the discussion on the old forum is that one can choose for a **fixed** zoom/window combo in any **fixed** size one wants by coding it the right way, but that does not really solve my problem. Strangely enough when I export a restricted instance as a standalone app or .exe, the panel behaves exactly the way I want.

The ideal situation would be to zoom in/out and the window resizes automatically to the size of the content. 

But maybe I should decide which fixed size (or maybe an extra size, like @damiensellier does) is the best for my panel as VST and/or AU plugin and set that size with the appropriate coding... OTOH I will have a look at the CTRLR source and who knows, I find the origin of the issue!

**Reply by damiensellier** [reply]
From what I understand some size properties only applies to the standalone apps or pannels opened in the ctrlr app. The solution would be to find the properties that resize the container when exported as VST/AU instance.
I'm pretty sure there's a function for that in JUCE, [they talk about it on the JUCE forum](https://forum.juce.com/t/best-way-to-implement-resizable-plugin/12644/8) and also [here](https://forum.juce.com/t/resizable-plugin-how-to/45043/6). 
They use "ResizableCornerComponent" for VST2 and AU and "setResizable(true, true)" for VST3 and AUv2.
But is it available in Ctrlr, that's the question?

Some plugins on the market allow resizing but require restarting the host. It has something to do with the size value stored in the plugin global properties. 

Absolutely, if you find the reason why the frame is not resizing but the canvas does, that would help a lot because having to export 2x mac vst + 2x mac AU + 2x PC vst is a bummer. 
Have a nice day !

**Reply by bijlevel** [reply]
I did study the CTRLR source, but it is hopeless work to find out why the window of the AU plugin is not resizable. Maybe it is hidden deep down in JUCE anyway. So I'm inclined to go for a fixed plugin size with the code shown on the old forum.

**Reply by damiensellier** [reply]
@bijlevel It's in the [JUCE library, it's there](https://github.com/RomanKubiak/ctrlr/blob/master/JUCE/modules/juce_gui_basics/windows/juce_ResizableWindow.cpp), but I don't know how this framework is included in Ctrlr and if ever this function is available/connected in Ctrlr. 
I can find traces of "ResizableCornerComponent" or "setResizable" in the "Window" JUCE class, that's a good start. Now the question is if Ctrlr is capable of taking this feature in consideration and how?

https://docs.juce.com/master/classResizableWindow.html 

We have the zoom ratio working, if we have the setResizable(true, true) we can do it :)

**Reply by bijlevel** [reply]
OK, I will have another look, but I'm not optimistic (which I generally am :-)).

**Reply by damiensellier** [reply]
When I will be more available I'll try to find a way to set the property via LUA so as to be able to drag the corner for resize. But it will require exporting vst instances from the ctrlr instigator plugin numerous times, until it's working as desired.

EDIT: I tried something but those function are null

```
	panel:getCanvas():setResizable(true, true)
	panel:getCanvas():setResizeLimits(1480, 1220, 2960, 2440)
```

**Reply by bijlevel** [reply]
For what it is worth, I did find these code lines in the source file CtrlrEditor.cpp:


`if (!JUCEApplication::isStandaloneApp())`

	{
	    if (owner.getInstanceMode() != InstanceSingleRestriced)
        {
            /* Restricted instances don't get to be resizable */
            addAndMakeVisible (&resizer);
            resizer.setAlwaysOnTop (false);
            resizer.grabKeyboardFocus();
            resizer.toFront (true);
        }
	}

So I believe the decision is made that restricted instances are not resizable... BTW this is the source of the latest version 5.6.23!

**Reply by damiensellier** [reply]
I don't know if it's the same for VST/AU instances but it definitely says it's locked unfortunately. 
I think We should add a thread in the "Issues" section to ask for an option resizeable (yes/no).

**Reply by bijlevel** [reply]
Always good to open an issue, but as you can see there are many, many issues not yet addressed, so I doubt if the author (Roman Kubiak or someone else) will ever find the time (or is it motivation?) to solve these

**Reply by bijlevel** [reply]
I did succeed in compiling v5.3.201 for Mac with a resizable window on restricted instance. This is only the standalone version. I did not succeed (and the way things look now, I never will!) in compiling the AU component for this version. The standalone version does also solve the problem of not listing the available resources, like v5.3.198 did.

The rather good news is that I did successful compile v5.6.23 (the latest published source) for Mac with the resizable window, including the AU component. So now the component has a resizable window. I did use this version for further development and it seems rather stable. The most important "change" is that you have to adjust the format of the (sysex) midi message send to the hardware. You can find more information about this in issue: https://github.com/RomanKubiak/ctrlr/issues/260#issue-860395933. Also I believe the "adding 24 pixels"-issue has been solved. I'm rather happy with this version, because it looks quite stable and I think  under the hood several issues have been solved. But of course we always have to keep our fingers crossed ;-)!

**Reply by damiensellier** [reply]
That's a very good news! And how do you manage to get the frame resizeable (locked/unlocked)? Do you define a global property with LUA or is it an option to tick in panel mode? 
Thanks

**Reply by bijlevel** [reply]
No, I adjusted the code in the source and compiled the application (v5.6.23) again. I did the same for v5.3.201, but unfortunately only the standalone app compiled OK, but not the AU component. The latter turned out to be a nightmare and I probably will give up trying it.

**Reply by damiensellier** [reply]
Oh alright so did you change false to true here ?

```
if (!JUCEApplication::isStandaloneApp())

{
    if (owner.getInstanceMode() != InstanceSingleRestriced)
    {
        /* Restricted instances don't get to be resizable */
        addAndMakeVisible (&resizer);
        resizer.setAlwaysOnTop (true);  /* false to true ??? */
        resizer.grabKeyboardFocus();
        resizer.toFront (true);
    }
}

```

**Reply by bijlevel** [reply]
No, I just commented out the check for restricted instance:

`if (!JUCEApplication::isStandaloneApp())

	{
	    //if (owner.getInstanceMode() != InstanceSingleRestriced)
        //{
            /* Restricted instances don't get to be resizable */
            addAndMakeVisible (&resizer);
            resizer.setAlwaysOnTop (false);
            resizer.grabKeyboardFocus();
            resizer.toFront (true);
        //}
	}`

and recompiled.

**Reply by damiensellier** [reply]
Well played. Thanks for the info ;)
